src/main/java/net/neoforged/moddevgradle/tasks/CreateMinecraftJar.java

```diff
@@ -0,0 +1,234 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.zip.ZipFile;
+
+/**
+ * The primary task for creating the Minecraft artifacts that mods will be compiled against,
+ * using the NFRT CLI.
+ */
+@ApiStatus.Experimental
+@ApiStatus.NonExtendable
+public abstract class CreateMinecraftJar extends DefaultTask {
+    @Inject
+    public CreateMinecraftJar() {
+    }
+
+    /**
+     * The installertools Jar needed to create the Minecraft jar.
+     */
+    @Classpath
+    public abstract ConfigurableFileCollection getInstallerTools();
+
+    /**
+     * Files added to this collection will be applied to the Minecraft jar.
+     */
+    @InputFiles
+    public abstract ConfigurableFileCollection getAccessTransformers();
+
+    /**
+     * Interface injection data files to apply to the Minecraft jar.
+     */
+    @InputFiles
+    public abstract ConfigurableFileCollection getInterfaceInjectionData();
+
+    /**
+     * This must contain a single file, which has one of the following supported formats:
+     * <ul>
+     * <li>NeoForge Userdev jar file, containing a Userdev config.json file</li>
+     * </ul>
+     */
+    @Classpath
+    public abstract ConfigurableFileCollection getNeoForgeUserDevConfig();
+
+    /**
+     * This must contain a single file, which has one of the following supported formats:
+     * <ul>
+     * <li>NeoForm config.json file.</li>
+     * <li>NeoForm data zip file.</li>
+     * </ul>
+     */
+    @Classpath
+    public abstract RegularFileProperty getNeoFormConfig();
+
+    /**
+     * The Minecraft jar.
+     */
+    @OutputFile
+    public abstract RegularFileProperty getMinecraftJar();
+
+    @Inject
+    @ApiStatus.Internal
+    protected abstract ExecOperations getExecOperations();
+
+    @TaskAction
+    public void createArtifacts() throws IOException {
+
+        File neoFormConfig = getNeoFormConfig().getAsFile().get();
+
+        // Parse the neoFormConfig JSON and extract the minecraft field
+        String minecraftVersion;
+        try (var neoFormZip = new ZipFile(neoFormConfig)) {
+            var configEntry = neoFormZip.getEntry("config.json");
+            if (configEntry == null) {
+                throw new InvalidUserCodeException("NeoForm config file does not contain a config.json entry.");
+            }
+            try (var inputStream = neoFormZip.getInputStream(configEntry);
+                 var reader = new InputStreamReader(inputStream, StandardCharsets.UTF_8)) {
+                Gson gson = new Gson();
+                JsonObject jsonObject = gson.fromJson(reader, JsonObject.class);
+                minecraftVersion = jsonObject.get("version").getAsString();
+            }
+        }
+
+        HttpClient client = HttpClient.newHttpClient();
+        Gson gson = new Gson();
+
+        // Download version manifest
+        HttpRequest manifestRequest = HttpRequest.newBuilder()
+                .uri(URI.create("https://piston-meta.mojang.com/mc/game/version_manifest_v2.json"))
```

Comment:
It would be great to cache this. Imagine you are in a train and suddenly Gradle decides to re-run this task for whatever reason. That might be an argument to use NFRT :P